atoma-proxy/src/server/handlers/chat_completions.rs

```diff
@@ -795,33 +808,70 @@ impl RequestModel for RequestModelChatCompletions {
 
         for message in &self.messages {
             let content = message
-                .get("content")
-                .and_then(|content| content.as_str())
+                .get(CONTENT_KEY)
+                .and_then(|content| MessageContent::deserialize(content).ok())
                 .ok_or_else(|| AtomaProxyError::RequestError {
                     message: "Missing or invalid message content".to_string(),
                     endpoint: CHAT_COMPLETIONS_PATH.to_string(),
                 })?;
 
-            let num_tokens = tokenizer
-                .encode(content, true)
-                .map_err(|err| AtomaProxyError::InternalError {
-                    message: format!("Failed to encode message: {err:?}"),
-                    endpoint: CHAT_COMPLETIONS_PATH.to_string(),
-                })?
-                .get_ids()
-                .len() as u64;
-
-            total_num_tokens += num_tokens;
-            // add 2 tokens as a safety margin, for start and end message delimiters
-            total_num_tokens += 2;
-            // add 1 token as a safety margin, for the role name of the message
-            total_num_tokens += 1;
+            match content {
+                MessageContent::Text(text) => {
+                    let num_tokens = tokenizer
+                        .encode(text, true)
+                        .map_err(|err| AtomaProxyError::InternalError {
+                            message: format!("Failed to encode message: {err:?}"),
+                            endpoint: CHAT_COMPLETIONS_PATH.to_string(),
+                        })?
+                        .get_ids()
+                        .len() as u64;
+
+                    total_num_tokens += num_tokens;
+                    // add 2 tokens as a safety margin, for start and end message delimiters
+                    total_num_tokens += 2;
+                    // add 1 token as a safety margin, for the role name of the message
+                    total_num_tokens += 1;
+                    // add the max completion tokens, to account for the response
+                    total_num_tokens += self.max_completion_tokens;
+                }
+                MessageContent::Array(parts) => {
+                    for part in parts {
+                        match part {
+                            MessageContentPart::Text { text, .. } => {
```

Comment:
if you have a conversation with 3 messages, and `max_completion_tokens` is 1000, you're adding:

Message 1: +1000 tokens
Message 2: +1000 tokens
Message 3: +1000 tokens

So you end up estimating 3000 completion tokens, is that correct?